test/realizations/catchments/Bmi_Cpp_Adapter_Test.cpp

```diff
@@ -0,0 +1,664 @@
+                                                utils::StreamHandler());
+    }
+    catch (const std::exception &e) {
+        std::clog << e.what() << std::endl;
+        throw e;
+    }
+ }
+
+void Bmi_Cpp_Adapter_Test::TearDown() {
+
+}
+
+std::string
+Bmi_Cpp_Adapter_Test::file_search(const std::vector<std::string> &parent_dir_options, const std::string& file_basename) {
+    // Build vector of names by building combinations of the path and basename options
+    std::vector<std::string> name_combinations;
+
+    // Build so that all path names are tried for given basename before trying a different basename option
+    for (auto & path_option : parent_dir_options)
+        name_combinations.push_back(path_option + file_basename);
+
+    return utils::FileChecker::find_first_readable(name_combinations);
+}
+
+/** Simple test to make sure the model initializes. */
+TEST_F(Bmi_Cpp_Adapter_Test, Initialize_0_a) {
+    adapter->Initialize();
+    adapter->Finalize();
+}
+
+/** Test output variables can be retrieved. */
+TEST_F(Bmi_Cpp_Adapter_Test, GetOutputVarNames_0_a) {
+    try {
+        ASSERT_EQ(adapter->GetOutputVarNames(), expected_output_var_names);
+    }
+    catch (std::exception& e) {
+        printf("Exception getting output var names: %s", e.what());
+        throw e;
+    }
+}
+
+/** Test output variables item count can be retrieved. */
+TEST_F(Bmi_Cpp_Adapter_Test, GetOutputItemCount_0_a) {
+    try {
+        ASSERT_EQ(adapter->GetOutputItemCount(), expected_output_var_names.size());
+    }
+    catch (std::exception& e) {
+        printf("Exception getting output var count: %s", e.what());
+        throw e;
+    }
+}
+
+/** Test that both the get value function works for input 1. */
+TEST_F(Bmi_Cpp_Adapter_Test, GetValue_0_a) {
+    adapter->Initialize();
+    double value = 5.0;
+    bmi::Bmi* model = friend_get_model_class(adapter.get());
+    //*model->input_var_1 = value;
+    model->SetValue("INPUT_VAR_1", &value);
+    double retrieved = adapter->GetValue<double>("INPUT_VAR_1")[0];
+    adapter->Finalize();
+    ASSERT_EQ(value, retrieved);
+}
+
+/** Test that both the get value function works for input 2. */
+TEST_F(Bmi_Cpp_Adapter_Test, GetValue_0_b) {
+    adapter->Initialize();
+    double value = 6.0;
+    bmi::Bmi* model = friend_get_model_class(adapter.get());
+    //*model->output_var_1 = value;
+    model->SetValue("INPUT_VAR_2", &value);
+    double retrieved = adapter->GetValue<double>("INPUT_VAR_2")[0];
+    adapter->Finalize();
+    ASSERT_EQ(value, retrieved);
+}
+
+/** Test that both the get value function works for output 1. */
+TEST_F(Bmi_Cpp_Adapter_Test, GetValue_0_c) {
+    adapter->Initialize();
+    double value = 7.0;
+    bmi::Bmi* model = friend_get_model_class(adapter.get());
+    //*model->output_var_1 = value;
+    model->SetValue("OUTPUT_VAR_1", &value);
+    double retrieved = adapter->GetValue<double>("OUTPUT_VAR_1")[0];
```

Comment:
Concur... This would be best as a refactor across all adapter tests, however. These tests deliberately mirror the C adapter tests....which also only test (support??) double values ATM. And I don't think anything that _uses_ the adapters supports anything other than double, yet.